### PR TITLE
Updates to #236 to avoid breaking change to `pybop.Optimisation`

### PR DESCRIPTION
**Pull request by BradyPlanden**
# Description

This PR applies changes to #236 so that v24.5 does not have breaking changes in the `pybop.Optimisation` API.

## Issue reference
Fixes # N/A

## Review
Before you mark your PR as ready for review, please ensure that you've considered the following:
- Updated the [CHANGELOG.md](https://github.com/pybop-team/PyBOP/blob/develop/CHANGELOG.md) in reverse chronological order (newest at the top) with a concise description of the changes, including the PR number.
- Noted any breaking changes, including details on how it might impact existing functionality.

## Type of change
- [ ] New Feature: A non-breaking change that adds new functionality.
- [ ] Optimization: A code change that improves performance.
- [ ] Examples: A change to existing or additional examples.
- [X] Bug Fix: A non-breaking change that addresses an issue.
- [ ] Documentation: Updates to documentation or new documentation for new features.
- [X] Refactoring: Non-functional changes that improve the codebase.
- [ ] Style: Non-functional changes related to code style (formatting, naming, etc).
- [ ] Testing: Additional tests to improve coverage or confirm functionality.
- [ ] Other: (Insert description of change)

# Key checklist:

- [X] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybop-team/PyBOP/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [X] All unit tests pass: `$ nox -s tests`
- [X] The documentation builds: `$ nox -s doctest`

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:
- [ ] Code is well-commented, especially in complex or unclear areas.
- [ ] Added tests that prove my fix is effective or that my feature works.
- [ ] Checked that coverage remains or improves, and added tests if necessary to maintain or increase coverage.

Thank you for contributing to our project! Your efforts help us to deliver great software.

### CODE REVIEW

File: examples/notebooks/optimiser_interface.ipynb
Comment (NicolaCourtier) [inline examples/notebooks/optimiser_interface.ipynb:1]: I'm not sure we need a dedicated notebook for this, unless we are planning to keep the Optimisation wrapper class long-term (to be discussed).

File: examples/scripts/spm_pso.py [hunk 1]
```diff
@@ -37,7 +37,7 @@
 # Generate problem, cost function, and optimisation class
 problem = pybop.FittingProblem(model, parameters, dataset)
 cost = pybop.SumSquaredError(problem)
-optim = pybop.PSO(cost, max_iterations=100)
+optim = pybop.Optimisation(cost, optimiser=pybop.SciPyMinimize, max_iterations=100)
```
Comment (NicolaCourtier) [inline examples/scripts/spm_pso.py:40]: Should be pybop.PSO

File: pybop/optimisers/_optimisation.py [hunk 2]
```diff
@@ -245,3 +245,51 @@ def set_allow_infeasible_solutions(self, allow=True):
...
+    allow_infeasible_solutions : bool, optional
+        If True, infeasible parameter values will be allowed in the optimisation (default: True).
+
+    Attributes
+    ----------
+    All attributes from the pybop.optimiser() class
+
+    """
+
+    def __init__(self, cost, optimiser=None, **optimiser_kwargs):
```
Comment (NicolaCourtier) [inline pybop/optimisers/_optimisation.py:280]: Perhaps add a deprecation warning here.

**Comment by BradyPlanden** [comment]
Ugh, the recent file rename messed with the number of changes. Happy to revert it to help with review, it's needed to remove circular import issue in the `Optimisation` class. @NicolaCourtier

**Comment by NicolaCourtier** [comment]
Yes please do revert it @BradyPlanden, let's change the filenames at the end once both PRs are ready.

File: examples/notebooks/optimiser_interface.ipynb [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3ef5b0da-f755-43c6-8904-79d7ee0f218c",
+   "metadata": {},
+   "source": [
+    "## Interacting with the Optimisation classes\n",
+    "\n",
+    "Next, we construct the optimisation class with our algorithm of choice and run it. For the first case, we show the direct optimiser interface and construct the `pybop.CMAES` method. For the sake of reducing the runtime of this example, we limit the maximum iterations to 50; however, feel free to update this value. To showcase the different methods to set optimisation options, this example sets the maximum iterations three different ways."
```
Comment (NicolaCourtier) [inline examples/notebooks/optimiser_interface.ipynb:342]: *two different ways

File: examples/notebooks/optimiser_interface.ipynb [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+     "iopub.execute_input": "2024-04-14T18:57:46.512725Z",
+     "iopub.status.busy": "2024-04-14T18:57:46.512597Z",
+     "iopub.status.idle": "2024-04-14T18:57:49.259154Z",
+     "shell.execute_reply": "2024-04-14T18:57:49.257712Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "optim_interface_one = pybop.CMAES(cost, max_iterations=50)\n",
+    "optim_interface_one.set_max_iterations(50)\n",
```
Comment (NicolaCourtier) [inline examples/notebooks/optimiser_interface.ipynb:360]: This line is not necessary and will only work for PINTS optimisers.

File: examples/notebooks/optimiser_interface.ipynb [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "479fc846",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "optim_interface_two = pybop.Optimisation(cost, optimiser=pybop.CMAES, max_iterations=50)\n",
+    "optim_interface_two.set_max_iterations(50)\n",
```
Comment (NicolaCourtier) [inline examples/notebooks/optimiser_interface.ipynb:380]: As above

File: tests/integration/test_parameterisations.py [hunk 6]
```diff
@@ -107,7 +107,12 @@ def test_spm_optimisers(self, optimiser, spm_costs):
             spm_costs.bounds = bounds
 
         # Test each optimiser
-        parameterisation = optimiser(cost=spm_costs, sigma0=0.05, max_iterations=125)
+        if optimiser in [pybop.PSO]:
+            parameterisation = pybop.Optimisation(
+                cost=spm_costs, optimiser=optimiser, sigma0=0.05
```
Comment (NicolaCourtier) [inline tests/integration/test_parameterisations.py:112]: Keep the max_iterations=125 setting?

File: tests/integration/test_parameterisations.py [hunk 7]
```diff
@@ -107,7 +107,12 @@ def test_spm_optimisers(self, optimiser, spm_costs):
...
+            )
+        else:
+            parameterisation = optimiser(cost=spm_costs, sigma0=0.05)
```
Comment (NicolaCourtier) [inline tests/integration/test_parameterisations.py:115]: As above

File: pybop/optimisers/optimisation.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+            self.optimiser = XNES(cost, **optimiser_kwargs)
+        elif issubclass(optimiser, BasePintsOptimiser):
+            self.optimiser = optimiser(cost, **optimiser_kwargs)
+        elif issubclass(optimiser, BaseSciPyOptimiser):
+            self.optimiser = optimiser(cost, **optimiser_kwargs)
+        else:
+            raise ValueError("Unknown optimiser type")
+
+    def run(self, **optimiser_kwargs):
+        return self.optimiser.run(**optimiser_kwargs)
```
Comment (NicolaCourtier) [inline pybop/optimisers/optimisation.py:48]: `optimiser_kwargs` can no longer be passed through run.

**Review by NicolaCourtier — APPROVED** [review]
Thanks @BradyPlanden for the addition of the `Optimisation` wrapper!

**Pull request merged**